File: backend/routes/intake.py
"""
API routes for user intake/profile normalization
"""
from fastapi import APIRouter, HTTPException, status, Body
from models.schemas import BaseResponse, ErrorResponse, IntakeResponse
from services.intake_service import IntakeService
from utils.security import sanitize_error_message
from typing import List, Optional, Dict, Any, Union
from pydantic import BaseModel, Field, field_validator

router = APIRouter()
intake_service = IntakeService()


class IntakeRequest(BaseModel):
    """Request schema for user intake/profile normalization"""
    skills: Optional[List[str]] = Field(
        None,
        description="List of skill names",
        min_length=0
    )
    interests: Optional[Union[List[str], str]] = Field(
        None,
        description="List of RIASEC interest categories (Realistic, Investigative, Artistic, Social, Enterprising, Conventional) or text description"
    )
    constraints: Optional[Dict[str, Any]] = Field(
        None,
        description="Dict with constraints: cost (min_wage), time (max_hours, flexible_hours), location (remote_preferred, location_preference)"
    )
    values: Optional[Dict[str, float]] = Field(
        None,
        description="Dict mapping career values to scores (0-7): impact, stability, flexibility"
    )
    
    @field_validator('skills')
    @classmethod
    def validate_skills(cls, v):
        if v is not None:
            # Remove empty strings and strip whitespace
            v = [s.strip() for s in v if s and s.strip()]
        return v
    
    @field_validator('interests')
    @classmethod
    def validate_interests(cls, v):
        if v is not None:
            if isinstance(v, str):
                # Text input - will be parsed later
                if not v.strip():
                    return None
                return v.strip()
            elif isinstance(v, list):
                # List input - validate RIASEC categories
                valid_categories = {
                    "Realistic", "Investigative", "Artistic",
                    "Social", "Enterprising", "Conventional"
                }
                normalized = []
                for item in v:
                    if isinstance(item, str) and item.strip():
                        item = item.strip()
                        # Case-insensitive matching
                        for cat in valid_categories:
                            if cat.lower() == item.lower():
                                normalized.append(cat)
                                break
                return normalized if normalized else None
        return v
    
    @field_validator('values')
    @classmethod
    def validate_values(cls, v):
        if v is not None:
            valid_keys = {"impact", "stability", "flexibility"}
            normalized = {}
            for key, value in v.items():
                key_lower = key.lower()
                # Find matching key (case-insensitive)
                for valid_key in valid_keys:
                    if valid_key.lower() == key_lower:
                        # Ensure value is in valid range (0-7)
                        if isinstance(value, (int, float)):
                            normalized[valid_key] = max(0.0, min(7.0, float(value)))
                        break
            return normalized if normalized else None
        return v
    
    @field_validator('constraints')
    @classmethod
    def validate_constraints(cls, v):
        if v is not None:
            normalized = {}
            # Cost constraint (min_wage)
            if "cost" in v:
                cost = v["cost"]
                if isinstance(cost, (int, float)) and cost >= 0:
                    normalized["min_wage"] = float(cost)
                elif isinstance(cost, dict) and "min_wage" in cost:
                    min_wage = cost["min_wage"]
                    if isinstance(min_wage, (int, float)) and min_wage >= 0:
                        normalized["min_wage"] = float(min_wage)
            
            # Time constraints
            if "time" in v:
                time_constraints = v["time"]
                if isinstance(time_constraints, dict):
                    if "max_hours" in time_constraints:
                        max_hours = time_constraints["max_hours"]
                        if isinstance(max_hours, (int, float)) and 0 < max_hours <= 168:
                            normalized["max_hours"] = float(max_hours)
                    if "flexible_hours" in time_constraints:
                        flexible = time_constraints["flexible_hours"]
                        if isinstance(flexible, bool):
                            normalized["flexible_hours"] = flexible
            
            # Location constraints
            if "location" in v:
                location_constraints = v["location"]
                if isinstance(location_constraints, dict):
                    if "remote_preferred" in location_constraints:
                        remote = location_constraints["remote_preferred"]
                        if isinstance(remote, bool):
                            normalized["remote_preferred"] = remote
                    if "location_preference" in location_constraints:
                        loc = location_constraints["location_preference"]
                        if isinstance(loc, str) and loc.strip():
                            normalized["location_preference"] = loc.strip()
            
            # Support legacy format for backward compatibility
            if "min_wage" in v:
                min_wage = v["min_wage"]
                if isinstance(min_wage, (int, float)) and min_wage >= 0:
                    normalized["min_wage"] = float(min_wage)
            if "remote_preferred" in v:
                remote = v["remote_preferred"]
                if isinstance(remote, bool):
                    normalized["remote_preferred"] = remote
            if "max_education_level" in v:
                edu_level = v["max_education_level"]
                if isinstance(edu_level, (int, float)) and 0 <= edu_level <= 5:
                    normalized["max_education_level"] = int(edu_level)
            
            return normalized if normalized else None
        return v


@router.post("/intake", response_model=BaseResponse)
async def process_intake(request: IntakeRequest = Body(...)):
    """
    Process user intake and return normalized profile + derived features summary
    
    Validates and normalizes:
    - skills: List of skill names
    - interests: RIASEC categories (list) or text description
    - constraints: cost (min_wage), time (max_hours, flexible_hours), location (remote_preferred, location_preference)
    - values: impact, stability, flexibility (0-7 scale)
    
    Returns normalized profile with feature vectors and summary statistics
    """
    try:
        result = intake_service.normalize_profile(
            skills=request.skills,
            interests=request.interests,
            constraints=request.constraints,
            values=request.values
        )
        
        return BaseResponse(
            success=True,
            message="Profile normalized successfully",
            data=result
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=ErrorResponse(
                success=False,
                message="Validation error",
                error=sanitize_error_message(e)
            ).model_dump()
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to process intake",
                error=sanitize_error_message(e)
            ).model_dump()
        )

File: frontend/src/services/intake.ts
/**
 * Intake Service
 * Service for normalizing user profiles and validating intake forms
 */
import apiClient from './apiClient';
import { BaseResponse } from './types';

/**
 * RIASEC interest categories
 */
export const RIASEC_CATEGORIES = [
  'Realistic',
  'Investigative',
  'Artistic',
  'Social',
  'Enterprising',
  'Conventional',
] as const;

export type RIASECCategory = typeof RIASEC_CATEGORIES[number];

/**
 * Career values
 */
export const CAREER_VALUES = ['impact', 'stability', 'flexibility'] as const;

export type CareerValue = typeof CAREER_VALUES[number];

/**
 * Intake request types
 */
export interface IntakeConstraints {
  cost?: number | { min_wage?: number };
  time?: {
    max_hours?: number;
    flexible_hours?: boolean;
  };
  location?: {
    remote_preferred?: boolean;
    location_preference?: string;
  };
  // Legacy format support
  min_wage?: number;
  remote_preferred?: boolean;
  max_education_level?: number;
}

export interface IntakeValues {
  impact?: number;
  stability?: number;
  flexibility?: number;
}

export interface IntakeRequest {
  skills?: string[];
  interests?: string[] | string; // RIASEC categories or text description
  constraints?: IntakeConstraints;
  values?: IntakeValues;
}

/**
 * Feature vectors in normalized profile
 */
export interface FeatureVectors {
  skill_vector: number[];
  interest_vector: number[];
  values_vector: number[];
  constraint_features: number[];
  combined_vector: number[];
}

/**
 * Interests structure in normalized profile
 */
export interface NormalizedInterests {
  riasec_scores: Record<string, number>;
  raw_input: string[] | string | null;
}

/**
 * Normalized profile structure (matches backend response)
 */
export interface NormalizedProfileData {
  skills: string[];
  interests: NormalizedInterests;
  values: Record<string, number>;
  constraints: Record<string, any>;
  feature_vectors: FeatureVectors;
}

/**
 * Dominant interest/value in summary
 */
export interface DominantItem {
  category?: string;
  value?: string;
  score: number;
}

/**
 * Feature vector statistics
 */
export interface FeatureVectorStats {
  dimension: number;
  non_zero_count: number;
  max_value: number;
  mean_value: number;
  std_value: number;
}

/**
 * Constraints summary
 */
export interface ConstraintsSummary {
  has_wage_constraint: boolean;
  has_location_constraint: boolean;
  has_time_constraint: boolean;
  constraint_count: number;
}

/**
 * Derived features summary (matches backend response)
 */
export interface DerivedFeaturesSummary {
  profile_completeness: number;
  dominant_interests: DominantItem[];
  dominant_values: DominantItem[];
  feature_vector_stats: FeatureVectorStats;
  constraints_summary: ConstraintsSummary;
}

/**
 * Complete intake response structure (matches backend response)
 */
export interface IntakeResponse {
  normalized_profile: NormalizedProfileData;
  derived_features_summary: DerivedFeaturesSummary;
}

/**
 * Normalized profile response (for backward compatibility)
 * @deprecated Use IntakeResponse instead
 */
export interface NormalizedProfile extends IntakeResponse {
  [key: string]: any;
}

/**
 * Validation errors
 */
export interface ValidationErrors {
  skills?: string;
  interests?: string;
  constraints?: string;
  values?: string;
  [key: string]: string | undefined;
}

/**
 * Validate skills input
 * @param skills - List of skill strings
 * @returns Error message if invalid, undefined if valid
 */
export const validateSkills = (skills?: string[]): string | undefined => {
  if (!skills || skills.length === 0) {
    return undefined; // Skills are optional
  }

  // Check if all items are non-empty strings
  const invalidSkills = skills.filter(
    (skill) => !skill || typeof skill !== 'string' || !skill.trim()
  );

  if (invalidSkills.length > 0) {
    return 'All skills must be non-empty strings';
  }

  // Check for reasonable length
  const tooLongSkills = skills.filter((skill) => skill.length > 100);
  if (tooLongSkills.length > 0) {
    return 'Skills must be 100 characters or less';
  }

  return undefined;
};

/**
 * Validate interests input
 * @param interests - RIASEC categories (list) or text description
 * @returns Error message if invalid, undefined if valid
 */
export const validateInterests = (
  interests?: string[] | string
): string | undefined => {
  if (!interests) {
    return undefined; // Interests are optional
  }

  if (typeof interests === 'string') {
    // Text description - validate length
    if (interests.trim().length === 0) {
      return 'Interest description cannot be empty';
    }
    if (interests.length > 500) {
      return 'Interest description must be 500 characters or less';
    }
    return undefined;
  }

  if (Array.isArray(interests)) {
    if (interests.length === 0) {
      return undefined; // Empty array is valid
    }

    // Validate RIASEC categories
    const validCategories = new Set(
      RIASEC_CATEGORIES.map((cat) => cat.toLowerCase())
    );
    const invalidCategories = interests.filter(
      (interest) =>
        !interest ||
        typeof interest !== 'string' ||
        !validCategories.has(interest.trim().toLowerCase())
    );

    if (invalidCategories.length > 0) {
      return `Invalid RIASEC categories. Valid categories are: ${RIASEC_CATEGORIES.join(', ')}`;
    }

    return undefined;
  }

  return 'Interests must be a string or an array of strings';
};

/**
 * Validate constraints input
 * @param constraints - Constraints object
 * @returns Error message if invalid, undefined if valid
 */
export const validateConstraints = (
  constraints?: IntakeConstraints
): string | undefined => {
  if (!constraints) {
    return undefined; // Constraints are optional
  }

  // Validate cost constraint
  if (constraints.cost !== undefined) {
    if (typeof constraints.cost === 'number') {
      if (constraints.cost < 0) {
        return 'Cost (min_wage) must be non-negative';
      }
    } else if (typeof constraints.cost === 'object') {
      if (
        constraints.cost.min_wage !== undefined &&
        (typeof constraints.cost.min_wage !== 'number' ||
          constraints.cost.min_wage < 0)
      ) {
        return 'min_wage must be a non-negative number';
      }
    } else {
      return 'Cost constraint must be a number or object with min_wage';
    }
  }

  // Validate time constraints
  if (constraints.time) {
    if (
      constraints.time.max_hours !== undefined &&
      (typeof constraints.time.max_hours !== 'number' ||
        constraints.time.max_hours <= 0 ||
        constraints.time.max_hours > 168)
    ) {
      return 'max_hours must be a number between 0 and 168';
    }

    if (
      constraints.time.flexible_hours !== undefined &&
      typeof constraints.time.flexible_hours !== 'boolean'
    ) {
      return 'flexible_hours must be a boolean';
    }
  }

  // Validate location constraints
  if (constraints.location) {
    if (
      constraints.location.remote_preferred !== undefined &&
      typeof constraints.location.remote_preferred !== 'boolean'
    ) {
      return 'remote_preferred must be a boolean';
    }

    if (
      constraints.location.location_preference !== undefined &&
      (typeof constraints.location.location_preference !== 'string' ||
        constraints.location.location_preference.trim().length === 0)
    ) {
      return 'location_preference must be a non-empty string';
    }
  }

  // Validate legacy format
  if (
    constraints.min_wage !== undefined &&
    (typeof constraints.min_wage !== 'number' || constraints.min_wage < 0)
  ) {
    return 'min_wage must be a non-negative number';
  }

  if (
    constraints.remote_preferred !== undefined &&
    typeof constraints.remote_preferred !== 'boolean'
  ) {
    return 'remote_preferred must be a boolean';
  }

  if (
    constraints.max_education_level !== undefined &&
    (typeof constraints.max_education_level !== 'number' ||
      constraints.max_education_level < 0 ||
      constraints.max_education_level > 5)
  ) {
    return 'max_education_level must be a number between 0 and 5';
  }

  return undefined;
};

/**
 * Validate values input
 * @param values - Values object with impact, stability, flexibility (0-7 scale)
 * @returns Error message if invalid, undefined if valid
 */
export const validateValues = (values?: IntakeValues): string | undefined => {
  if (!values) {
    return undefined; // Values are optional
  }

  const validKeys = new Set(CAREER_VALUES);
  const valueRange = { min: 0, max: 7 };

  for (const [key, value] of Object.entries(values)) {
    if (!validKeys.has(key as CareerValue)) {
      return `Invalid value key: ${key}. Valid keys are: ${CAREER_VALUES.join(', ')}`;
    }

    if (typeof value !== 'number') {
      return `Value ${key} must be a number`;
    }

    if (value < valueRange.min || value > valueRange.max) {
      return `Value ${key} must be between ${valueRange.min} and ${valueRange.max}`;
    }
  }

  return undefined;
};

/**
 * Validate entire intake request
 * @param request - Intake request object
 * @returns Validation errors object (empty if valid)
 */
export const validateIntakeRequest = (
  request: IntakeRequest
): ValidationErrors => {
  const errors: ValidationErrors = {};

  const skillsError = validateSkills(request.skills);
  if (skillsError) {
    errors.skills = skillsError;
  }

  const interestsError = validateInterests(request.interests);
  if (interestsError) {
    errors.interests = interestsError;
  }

  const constraintsError = validateConstraints(request.constraints);
  if (constraintsError) {
    errors.constraints = constraintsError;
  }

  const valuesError = validateValues(request.values);
  if (valuesError) {
    errors.values = valuesError;
  }

  return errors;
};

/**
 * POST /api/intake/intake - Normalize user profile
 * Validates and normalizes user profile data
 */
export const normalizeUserProfile = async (
  request: IntakeRequest
): Promise<IntakeResponse> => {
  // Validate request before sending
  const validationErrors = validateIntakeRequest(request);
  const hasErrors = Object.keys(validationErrors).length > 0;

  if (hasErrors) {
    throw new Error(
      `Validation failed: ${Object.values(validationErrors).join(', ')}`
    );
  }

  const response = await apiClient.post<BaseResponse<IntakeResponse>>(
    '/api/intake/intake',
    request
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to normalize user profile');
  }

  return response.data.data!;
};

/**
 * Intake service object with all methods
 */
export const intakeService = {
  normalizeUserProfile,
  validateSkills,
  validateInterests,
  validateConstraints,
  validateValues,
  validateIntakeRequest,
};

export default intakeService;

